stl/inc/vector

```diff
@@ -986,18 +1176,33 @@ public:
         const auto _Oldcapacity = static_cast<size_type>(_My_data._Myend - _Myfirst);
         if (_Newsize > _Oldcapacity) { // reallocate
             _Clear_and_reserve_geometric(_Newsize);
-            _Mylast = _Uninitialized_fill_n(_Myfirst, _Newsize, _Val, _Al);
+            if constexpr (is_nothrow_copy_constructible_v<_Ty>) {
+                _Mylast = _Uninitialized_fill_n(_Myfirst, _Newsize, _Val, _Al);
+                _ASAN_VECTOR_CREATE;
+            } else {

```

Comment:
So there is a bug here, we should pass `size_type` not `difference _type`

However, I cannot follow your argument here. We pass `_Mylast` to `_Extent_annotation`

Reply:
Note that here, we did call `_Clear_and_reserve_geometric` So we have fresh unannotated memory

Reply:
> So there is a bug here, we should pass `size_type` not `difference _type`

I'll audit all arguments to the macros, and ensure that (1) literals (`1`s) are passed simply/directly as `int` which may be narrowed to `difference_type` without diagnostics since they are constant expressions, (2) any explicit argument conversions are to `size_type` and not `difference_type`, (3) we don't redundantly explicitly convert arguments from `size_type` to `size_type`.

> However, I cannot follow your argument here. We pass `_Mylast` to `_Extent_annotation`

Yes, "fresh unannotated memory" contains no objects so `_Mylast == _Myfirst`.

Reply:
`__sanitizer_annotate_contiguous_container` does not check that the mid is correct. It's essentially a potential starting point of a memset.

Reply:
So how do we audit calls for correctness? I'll guess. Given a block of memory:
```
v- first                   end -v
|==============|----------------|
     old_last -^
```
there are two basic cases. (1) we want to extend access (`old_last < _new_last`):
```
v- first                   end -v
|==============|------|---------|
     old_last -^      ^- new_last
```
(2) we want to restrict access (`old_last > new_last`):
```
v- first                   end -v
|==============|======|---------|
     new_last -^      ^- old_last
```
I assume for (1), `__s_a_c_c` needs to fill `[old_last, new_last)` with `=`, and for (2) it needs to fill `[new_last, old_last)` with `-`. If that's the case, then the value we pass for `old_last` could be as low as `first` [for (1)] or as high as `end` [for (2)] at a cost of performance but with no impact to correctness since we would be redundantly setting some cells to the access they already had?

Do I also assume correctly that a freshly-allocated block of memory is fully accessible before the first call to `__s_a_c_c`, so that we'd ideally pass `end` for `old_last` in the first annotation call?

If those assumptions are correct, then the bottom branch here does the wrong thing by passing `first` as `old_last` via `_Extend_annotation` - `__s_a_c_c` will redundantly scribble the "allow access" value over `[first, new_last)` when all of `[first, end)` was already accessible (since it's a fresh block) leaving `[new_last, end)` incorrectly annotated as accessible.

Reply:
Yeah this is wrong. On success we should be calling `_ASAN_VECTOR_EXTEND`.

Reply:
As discussed in discord, the "cleanest" way forward in case of `_Clear_and_reserve_geometric` would be to catch exceptions and then call `_ASAN_VECTOR_CREATE` to invalidate the whole new memory range

Reply:
@cbezault This also indicates a lack of test coverage. We should add tests, that assign with a reallocation. something like
```cpp
    vec assign_size_count_reallocate(2, 0);
    assign_size_count_reallocate.assign(10, 1);
```

Same for the other cases